libs/eavmlib/src/ahttp_client.erl

```diff
@@ -0,0 +1,411 @@
+    {ok, Conn, closed};
+stream(#http_client{socket = {ssl, SSLSocket}, parser = Parser} = Conn, {ssl, SSLSocket, Chunk}) ->
+    {ok, UpdatedParser, Parsed} = feed_parser(Parser, Chunk),
+    Responses = make_responses(Parsed, Conn#http_client.ref, []),
+    {ok, Conn#http_client{parser = UpdatedParser}, Responses};
+stream(_Conn, _Other) ->
+    unknown.
+
+stream_data(#http_client{parser = Parser} = Conn, Chunk) ->
+    {ok, UpdatedParser, Parsed} = feed_parser(Parser, Chunk),
+    Responses = make_responses(Parsed, Conn#http_client.ref, []),
+    {ok, Conn#http_client{parser = UpdatedParser}, Responses}.
+
+make_responses([], _Ref, Acc) ->
+    Acc;
+make_responses([Response | Tail], Ref, Acc) when is_atom(Response) ->
+    make_responses(Tail, Ref, [{Response, Ref} | Acc]);
+make_responses([{Tag, Value} | Tail], Ref, Acc) ->
+    make_responses(Tail, Ref, [{Tag, Ref, Value} | Acc]).
+
+feed_parser(#parser_state{state = body} = Parser, Chunk) ->
+    consume_bytes(append_chunk(Parser, Chunk), []);
+feed_parser(Parser, Chunk) ->
+    consume_lines(append_chunk(Parser, Chunk), []).
+
+consume_bytes(#parser_state{acc = undefined} = Parser, ParsedAcc) ->
+    {ok, Parser, ParsedAcc};
+consume_bytes(#parser_state{acc = Chunk} = Parser, ParsedAcc) when is_binary(Chunk) ->
+    ReplacedAccParser = replace_chunk(Parser, undefined),
+    NewRemBodyBytes = maybe_decrement(
+        ReplacedAccParser#parser_state.remaining_body_bytes, byte_size(Chunk)
+    ),
+    {UpdatedParser, Parsed} = maybe_handle_end_of_response(
+        ReplacedAccParser#parser_state{remaining_body_bytes = NewRemBodyBytes}, [
+            {data, Chunk} | ParsedAcc
+        ]
+    ),
+    {ok, UpdatedParser, Parsed}.
+
+consume_lines(Parser, ParsedAcc) ->
+    case binary:split(Parser#parser_state.acc, <<"\r\n">>) of
+        [_NotTerminatedLine] ->
+            {ok, Parser, ParsedAcc};
+        [Line, Rest] ->
+            ReplacedAccParser = replace_chunk(Parser, Rest),
+            case parse_line(ReplacedAccParser, Line) of
+                {consume_bytes, UpdatedParser} ->
+                    consume_bytes(UpdatedParser, ParsedAcc);
+                {ok, UpdatedParser} ->
+                    consume_lines(UpdatedParser, ParsedAcc);
+                {ok, UpdatedParser, Found} ->
+                    consume_lines(UpdatedParser, [Found | ParsedAcc]);
+                {error, UpdatedParser, NotParsed} ->
+                    {error, UpdatedParser, ParsedAcc, NotParsed}
+            end
+    end.
+
+append_chunk(#parser_state{acc = undefined} = Parser, Chunk) ->
+    Parser#parser_state{acc = Chunk};
+append_chunk(#parser_state{acc = Acc} = Parser, Chunk) ->
+    Parser#parser_state{acc = <<Acc/binary, Chunk/binary>>}.
+
+replace_chunk(Parser, <<>>) ->
+    Parser#parser_state{acc = undefined};
+replace_chunk(Parser, Chunk) ->
+    Parser#parser_state{acc = Chunk}.
+
+parse_line(
+    #parser_state{state = undefined} = Parser, <<"HTTP/1.0 ", C:3/binary, " ", _Txt/binary>>
```

Comment:
We don't support HTTP 0.9 ? 😁

Reply:
I was going to go into the investigating http 0.9 rabbit hole, but then I just realized that is past midnight... :skull_and_crossbones: